### PR TITLE
feat: whitelist tf hooks

### PR DESCRIPTION
**Pull request by jcompagni10**
for https://github.com/neutron-org/neutron/pull/587 -- Allow chain manager dao to whitelist tokenfactory hooks

**Related PRs**
https://github.com/neutron-org/neutron-dao/pull/107
https://github.com/neutron-org/neutron-sdk/pull/150
https://github.com/neutron-org/neutronjsplus/pull/43

**Integration Test Run**
https://github.com/neutron-org/neutron-tests/actions/runs/9765613806

### CODE REVIEW

**Review by sotnikov-s — APPROVED** [review]

**Review by pr0n00gler — DISMISSED** [review]

File: src/testcases/parallel/tokenfactory.test.ts [hunk 1]
```diff
@@ -13,19 +14,70 @@ import {
...
+  neutronChain: CosmosWrapper,
+  subDao: Dao,
+  subdaoMember1: DaoMember,
+  codeID: number,
+  denomCreator: string,
+) {
+  const chainManagerAddress = (await neutronChain.getChainAdmins())[0];
+  const proposalId = await subdaoMember1.submitUpdateParamsTokenfactoryProposal(
+    chainManagerAddress,
+    'Proposal #2',
```
Comment (sotnikov-s) [inline src/testcases/parallel/tokenfactory.test.ts:43]: given the fact you call this func several times, it's incorrect (and doesn't actually have any point to do so) to set the `#2` number to the proposal in the title

File: src/testcases/parallel/tokenfactory.test.ts [hunk 2]
```diff
@@ -13,19 +14,70 @@ import {
...
+    'Cron update params proposal. Will pass',
```
Comment (sotnikov-s) [inline src/testcases/parallel/tokenfactory.test.ts:44]: it's not cron params update proposal. please give your changes a thorough look in the context of the change when copy+pasting

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

**Comment by jcompagni10** [comment]
https://github.com/neutron-org/neutron-tests/actions/runs/9670374932

**Review by sotnikov-s — COMMENTED** [review]
excuse me, one more thing. please add a test that asserts that hooks fail if not whitelisted

File: src/testcases/parallel/tokenfactory.test.ts [hunk 3]
```diff
@@ -346,59 +346,69 @@ describe('Neutron / Tokenfactory', () => {
...
-      }>(contractAddress, {
-        sudo_result_block_before: {},
+      test('set non-whitelisted hook fails  ', async () => {
+        const res = await msgSetBeforeSendHook(
+          neutronAccount,
+          ownerWallet.address.toString(),
+          newTokenDenom,
+          contractAddress,
+        );
+        expect(res.code).not.toEqual(0); // set hook fails
```
Comment (sotnikov-s) [inline src/testcases/parallel/tokenfactory.test.ts:356]: would be more accurate to check that the exact error is returned, not that just something went wrong

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

File: src/testcases/parallel/tokenfactory.test.ts [hunk 4]
```diff
@@ -197,6 +342,39 @@ describe('Neutron / Tokenfactory', () => {
...
+        'create_denom',
+        'new_token_denom',
+      );
+      const res2 = await msgSetBeforeSendHook(
+        neutronAccount,
+        ownerWallet.address.toString(),
+        newTokenDenom,
+        contractAddress,
+      );
+      expect(res2.code).toEqual(14); // "beforeSendHook is not whitelisted"
```
Comment (sotnikov-s) [inline src/testcases/parallel/tokenfactory.test.ts:361]: it's good enough to check the code, but just FYI for the future: a more convenient IMO way to check errors is by the error message e.g. like here (or actually in many other places in our tests) https://github.com/neutron-org/neutron-integration-tests/blob/601f0614b1c5e57dc7c75f91b4e20701fc4527ca/src/testcases/run_in_band/tge.auction.test.ts#L1977-L1988

File: src/testcases/parallel/tokenfactory.test.ts [hunk 5]
```diff
@@ -13,19 +14,69 @@ import {
...
+  );
+
+  let timelockedProp = await subdaoMember1.supportAndExecuteProposal(
+    proposalId,
+  );
+  await waitSeconds(10);
+  await subdaoMember1.executeTimelockedProposal(proposalId);
+  timelockedProp = await subDao.getTimelockedProposal(proposalId);
+  expect(timelockedProp.id).toEqual(proposalId);
+  expect(timelockedProp.status).toEqual('executed');
```
Comment (sotnikov-s) [inline src/testcases/parallel/tokenfactory.test.ts:66]: didn't this (speeded up?) approach work? I can see that you added it and then removed in the latest commits
image

**Review by sotnikov-s — DISMISSED** [review]
comments above are only informative, so approving it

File: src/testcases/parallel/tokenfactory.test.ts (hunk 4 above)
Comment (jcompagni10) [reply src/testcases/parallel/tokenfactory.test.ts:361]: could be mistaken but the way `msgSetBeforeSendHook` is called I can't use `rejects`

**Pull request merged**